File: argon-io/shared/src/main/scala/dev/argon/io/ResourceRecorder.scala
package dev.argon.io

import java.io.IOException
import zio.*

trait ResourceRecorder {
  def recordBinaryResource[R, E >: IOException](resource: BinaryResource[R, E]): ZIO[R, E, String]
  def recordDirectoryResource[R, E >: IOException](resource: DirectoryResource[R, E, BinaryResource]): ZIO[R, E, String]
}

File: plugin/api/java/src/main/java/dev/argon/plugin/api/ResourceRecorder.java
package dev.argon.plugin.api;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.math.BigInteger;

public interface ResourceRecorder<E extends Throwable> {
    @NotNull String recordBinaryResource(BinaryResource<E> resource) throws E, IOException, InterruptedException;
    @NotNull String recordDirectoryResource(DirectoryResource<E, BinaryResource<E>> resource) throws E, IOException, InterruptedException;
}
